### PR TITLE
Refactoing + avoid redundant re-building

### LINKED ISSUE #6: Do not execute build plan when the commit is already in history
Oftentimes we run a build for a commit that is already in history. This happens when between two Stackage HEAD builds no new commits appeared in GHC master. Obviously this sort of double-checking cannot provide us with any useful information (it may although give a different result if we have some packages with flaky test suites, but it's not valuable info anyway).

We should first check history file and skip execution of build plan for commits that are already in history .

### PR DESCRIPTION
**Pull request by mrkkrp**
Close #6.

### CODE REVIEW

File: Stackage/HEAD/History.hs [hunk 1]
```diff
@@ -1,90 +1,172 @@
...
+    let item = HistoryItem target sha1 buildUrl
+    -- Assert that reportPath is obtainable
+    void (reportPathMaybe item)
+    -- Assert that logPath is obtainable
+    void (logsPathMaybe item)
+    -- Assert that target is a valid RText 'PathPiece
+    void (URI.mkPathPiece target)
+    -- Assert that commit is a valid RText 'PathPiece
+    void (URI.mkPathPiece sha1)
+    return item
```
Comment (facundominguez) [inline Stackage/HEAD/History.hs]: The preconditions should be made explicit in the documentation of the smart constructor.

File: Stackage/HEAD/History.hs [hunk 2]
```diff
@@ -1,90 +1,172 @@
...
+
+-- | Create a 'HistoryItem' value.
+
+mkHistoryItem :: MonadThrow m
+  => Text              -- ^ Target name
+  -> Text              -- ^ Commit SHA1
+  -> URI               -- ^ Link to the build
+  -> m HistoryItem
+mkHistoryItem target sha1 buildUrl =
+  maybe (throwM $ InvalidHistoryItem target sha1 buildUrl) return $ do
```
Comment (facundominguez) [inline Stackage/HEAD/History.hs]: Using `Maybe` is going to yield a poor indication of what the problem was. It shouldn't be difficult to provide more information.

File: Stackage/HEAD/History.hs [hunk 3]
```diff
@@ -1,90 +1,172 @@
...
+
+hip :: HistoryItem -> Text
+hip (HistoryItem target sha1 _) = target <> "-" <> sha1
 
 -- | Convert 'HistoryItem' to a 'String'.
 
-strHistoryItem :: HistoryItem -> FilePath
-strHistoryItem = T.unpack . unHistoryItem
+ship :: HistoryItem -> String
+ship = T.unpack . hip
```
Comment (facundominguez) [inline Stackage/HEAD/History.hs]: No, no, no. Please use a longer name. `historyItemToString`, `showHistoryItem`?? If we are going to use some compression algorithm for function identifiers, that should be documented somewhere. I know there is also `hip`, but that name is also too short.

File: app/Main.hs [hunk 4]
```diff
@@ -265,6 +270,28 @@ generateSite siteDirRaw flakyPkgs reportsDir = do
   Site.generateSite params
   putStrLn "Done."
 
+-- | Check whether we've already seen this target\/commit combo. (If so, it
+-- makes sense for the CI script to skip building a build plan again).
```
Comment (facundominguez) [inline app/Main.hs]: Isn't this documentation misguiding? The function only checks if the newest item in the history has the same attributes (whatever they are) as the provided metadata info.

The documentation of the function and the command line argument need to be more precise.

File: Stackage/HEAD/Trac.hs [hunk 5]
```diff
@@ -29,6 +29,5 @@ generateTracTicket thisBuildUrl hitems buildDiff =
   "  * Execute `cabal new-build` or `cabal new-test` and see if you can reproduce the failure.\n\n" <>
   "The full CircleCI build log is available online here: " <> URI.render thisBuildUrl
   where
-    ghcCommit        = URI.render (ghcCommitUrl commit)
-    cabalConfig      = URI.render (cabalConfigUrl target)
-    (target, commit) = decomposeHistoryItem (snd hitems)
+    ghcCommit   = URI.render $ cabalConfigUrl (snd hitems)
+    cabalConfig = URI.render $ ghcCommitUrl   (snd hitems)
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply Stackage/HEAD/Trac.hs]: It does. Previously a `HistoryItem` was just a wrapper around `Text` which contained something like `nightly-2018-04-04-<ghc-commit-hash>.csv`. And so to get `target` (`nightly-2018-04-04`) and the commit we had to `decomposeHistoryItem` and then from that we could map to `URI`s to Cabal config and GHC commit at Phabricator. Now everything is nicely separated already in `HistoryItem` and its smart constructor guarantees that we can obtain those URLs purely.

File: app/Main.hs [hunk 6]
```diff
@@ -100,23 +91,35 @@ optionsParser = Options
...
+    targetOpt = (fmap T.pack . strOption . mconcat)
+      [ long "target"
+      , metavar "TARGET"
+      , help "Target name, without extension (lts-x.y, nightly-yyyy-mm-dd)"
+      ]
+    flakyPkgsOpt = toPackageNameSet <$> many ((strOption . mconcat)
+      [ long "flaky"
+      , metavar "PKG"
+      , help "Flaky package, changes in its state should always be considered innocent"
+      ])
```
[the remark this answers is not in the document]
Comment (mrkkrp) [reply app/Main.hs]: Sorry, I'll keep this in mind for future changes.

File: Stackage/HEAD/Trac.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrkkrp) [reply Stackage/HEAD/Trac.hs]: Ah sorry, indeed I swapped them!

File: Stackage/HEAD/History.hs (hunk 2 above)
Comment (mrkkrp) [reply Stackage/HEAD/History.hs]: Generally yes, but this should rarely fail if ever and it won't be difficult to see what's wrong by having `HistoryItemException` anyway.

File: Stackage/HEAD/History.hs (hunk 3 above)
Comment (mrkkrp) [reply Stackage/HEAD/History.hs]: Okay, fair point.

**Pull request merged**